### PR TITLE
Skip lazy imports when the caller is inspect.py

### PR DESCRIPTION
**Pull request by asumagic**
## What does this PR do?

This avoids having certain inspect functions import our lazy modules when we don't want them to. `getframeinfo` in particular appears to do it, and this gets called by PyTorch at some point. IPython might also be doing it but autocomplete still seems to work.

This does not appear to break anything. Added test for hyperpyyaml to ensure we're not breaking that.

Fixes the issue where the redirection warning was emitted even if you weren't using `speechbrain.pretrained`, and prevents some lazy imports from forcefully being imported.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Review by pplantinga — COMMENTED** [review]
I'll be honest, some of what's going on here is a bit beyond me without studying it more, but it seems to work.

Just a question though, I get wanting to prevent imports that are unnecessary but was it causing crashes or is it just to improve speed of loading etc.? Just asking because this adds some complexity to the lazy modules and I want to make sure its worth it.

**Comment by asumagic** [comment]
> I'll be honest, some of what's going on here is a bit beyond me without studying it more, but it seems to work.
> 
> Just a question though, I get wanting to prevent imports that are unnecessary but was it causing crashes or is it just to improve speed of loading etc.? Just asking because this adds some complexity to the lazy modules and I want to make sure its worth it.

The problem is that the `pretrained`->`inference` lazy import raises a warning at import time because of the issue I mentioned, so the warning would often get printed even if  you were never importing `.pretrained`.

Theoretically, we could instead have a check in the warning rather than doing this, but I believe that it could result in the warning never being raised in some cases (since the lazy module will no longer do anything once the import is done, so it couldn't catch a "genuine" `import speechbrain.pretrained` if `.pretrained` already got imported "accidentally").

File: speechbrain/utils/importutils.py [hunk 1]
```diff
@@ -43,9 +44,45 @@ def __init__(
...
+        code will inadvertently cause our modules to be imported, such as some
+        of PyTorch's op registering machinery.
+
+        Arguments
+        ---------
+        stacklevel : int
+            The stack trace level of the function that caused the import to
+            occur, relative to the **caller** of this function (e.g. if in
+            function `f` you call `_ensure_module(1)`, it will refer to the
+            function that called `f`).
```
Comment (Adel-Moumen) [inline speechbrain/utils/importutils.py:62]: could you please add a raise docstring as well e.g.: 
```
RAISES
--------
```
